### PR TITLE
Add DevOps metric Lead Time for Changes

### LINKED ISSUE #3501: Add DevOps metric Lead time for changes
Lead time for changes = average lead time of changes **completed** in a certain time period.

Rationale (why measure the lead time for changes):
- The shorter the lead time of changes, the sooner the new features can be used. Also, the shorter the lead time of changes, the fewer changes are in progress at the same time. Less context switching is needed and the risk of interfering changes is reduced.

Parameters:
- Number of days ago (time period)

Sources:
- Jira
- Azure DevOps

Timestamps to use:
- Start of the change:
  - In progress date
- End of the change:
  - Resolution date  

Filters:
- Types of issues, etc via a JQL (same as for the issues metric).

Out of scope:
- GitLab does not support the issues metric at the moment, so we're also not going to implement GitLab support for the lead time metric for now.

### PR DESCRIPTION
**Pull request by wkoot**
Closes #3501

### CODE REVIEW

File: components/collector/src/source_collectors/azure_devops/issues.py [hunk 1]
```diff
@@ -52,9 +79,20 @@ async def _parse_value(self, responses: SourceResponses) -> Value:
...
         """Return the work items, if any."""
-        return (await responses[1].json())["value"] if len(responses) > 1 else []
+        if len(responses) <= 1:  # the first call is to workItems, which only returns ids
+            return []
+        all_work_items = []
+        for response in responses[1:]:
+            response_json = (await response.json())
+            if not response_json:
+                continue
+            all_work_items.extend(response_json.get("value"))
```
Comment (fniessink) [inline components/collector/src/source_collectors/azure_devops/issues.py]: Why a guard clause if the work is only one line? How about this? Two fewer lines:
```suggestion
            if response_json := (await response.json()):
                all_work_items.extend(response_json.get("value"))
```

File: components/collector/src/source_collectors/azure_devops/lead_time_for_changes.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+from model import Entity, Entities, SourceResponses
+
+from .issues import AzureDevopsIssues
+
+
+class AzureDevopsLeadTimeForChanges(AzureDevopsIssues):
+    """Collector to calculate lead time for changes from Azure Devops Server."""
+
+    def _item_select_fields(self) -> list[str]:
+        """Also request date fields to calculate lead time."""
```
Comment (fniessink) [inline components/collector/src/source_collectors/azure_devops/lead_time_for_changes.py]:
```suggestion
        """Extend to also request date fields to calculate lead time."""
```

File: components/collector/src/source_collectors/azure_devops/lead_time_for_changes.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        base_fields = super()._item_select_fields()
+        return base_fields + ["System.CreatedDate", "System.ChangedDate"]
```
Comment (fniessink) [inline components/collector/src/source_collectors/azure_devops/lead_time_for_changes.py]: Is this so bad?

```suggestion
        return super()._item_select_fields() + ["System.CreatedDate", "System.ChangedDate"]
```

File: components/collector/src/source_collectors/azure_devops/lead_time_for_changes.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+        return days_ago(parse(finished_date)) <= int(cast(str, self._parameter("lookback_days")))
+
+    async def _parse_value(self, responses: SourceResponses) -> Value:
+        work_items = (await self._work_items(responses))
+        if not work_items:
+            return None
+
+        lead_times = [self.__lead_time(item) for item in work_items]
+        return str(round(mean(lead_times)))
```
Comment (fniessink) [inline components/collector/src/source_collectors/azure_devops/lead_time_for_changes.py]: I would do this. Maybe we need to decide how many lines of indented code warrants a guard clause ;-)

```suggestion
        if work_items := (await self._work_items(responses)):
            lead_times = [self.__lead_time(item) for item in work_items]
            return str(round(mean(lead_times)))
        return None
```

File: components/collector/src/source_collectors/azure_devops/lead_time_for_changes.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+    async def _parse_entities(self, responses: SourceResponses) -> Entities:
+        """Override to add the story points to the entities."""
```
Comment (fniessink) [inline components/collector/src/source_collectors/azure_devops/lead_time_for_changes.py]: Doc string seems wrong? Copy-paste-forget?

File: components/collector/src/source_collectors/azure_devops/user_story_points.py [hunk 6]
```diff
@@ -1,25 +1,39 @@
...
+from model import Entity, Entities, SourceResponses
 
 from .issues import AzureDevopsIssues
 
 
 class AzureDevopsUserStoryPoints(AzureDevopsIssues):
     """Collector to get user story points from Azure Devops Server."""
 
+    def _item_select_fields(self) -> list[str]:
+        """Also request date fields to calculate lead time."""
```
Comment (fniessink) [inline components/collector/src/source_collectors/azure_devops/user_story_points.py]: Doc string seems wrong.

File: components/collector/src/source_collectors/azure_devops/user_story_points.py [hunk 7]
```diff
@@ -1,25 +1,39 @@
...
+                key=work_item["id"],
+                project=work_item["fields"]["System.TeamProject"],
+                title=work_item["fields"]["System.Title"],
+                work_item_type=work_item["fields"]["System.WorkItemType"],
+                state=work_item["fields"]["System.State"],
+                url=work_item["_links"]["html"]["href"],
+                story_points=self.__story_points(work_item),
+            )
+            for work_item in await self._work_items(responses)
+        )
```
Comment (fniessink) [inline components/collector/src/source_collectors/azure_devops/user_story_points.py]: Lots of duplication with `AzureDevopsLeadTimeForChanges._parse_entities()`

File: components/collector/src/source_collectors/azure_devops/user_story_points.py [hunk 8]
```diff
@@ -1,25 +1,39 @@
...
 
     async def _parse_value(self, responses: SourceResponses) -> Value:
         """Override to parse the sum of the user story points from the responses."""
         calculated_value = sum(self.__story_points(work_item) for work_item in await self._work_items(responses))
-        return str(int(calculated_value))
+        return str(decimal_round_half_up(calculated_value))  # follow frontend behaviour
```
Comment (fniessink) [inline components/collector/src/source_collectors/azure_devops/user_story_points.py]: Rounding up floats should not need a comment really, maybe just remove it?

File: components/collector/src/source_collectors/jira/lead_time_for_changes.py [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+        if not finished_date:
+            return False
+
+        return days_ago(parse(finished_date)) <= int(cast(str, self._parameter("lookback_days")))
+
+    @classmethod
+    def _compute_value(cls, entities: Entities) -> Value:
+        if not entities:
+            return None
+
```
Comment (fniessink) [inline components/collector/src/source_collectors/jira/lead_time_for_changes.py]: Why is this necessary? Wouldn't (shouldn't) the rest of the method just work for an empty list of entities?

File: components/collector/tests/source_collectors/source_collector_test_case.py [hunk 10]
```diff
@@ -54,11 +56,13 @@ async def collect(
...
         )
-        post_response = self.__post_response(post_request_json_return_value)
+        post_response = self.__post_response(post_request_json_return_value, post_request_json_side_effect)
         get = AsyncMock(return_value=get_response, side_effect=get_request_side_effect)
         post = AsyncMock(return_value=post_response, side_effect=post_request_side_effect)
         with patch("aiohttp.ClientSession.get", get), patch("aiohttp.ClientSession.post", post):
             async with aiohttp.ClientSession() as session:
+                if return_mocks:
+                    return await MetricCollector(session, self.metric).collect(), get, post
                 return await MetricCollector(session, self.metric).collect()
```
Comment (fniessink) [inline components/collector/tests/source_collectors/source_collector_test_case.py]:
```suggestion
                result = await MetricCollector(session, self.metric).collect()
                return result, get, post if return_mocks else result
```

File: components/collector/tests/source_collectors/source_collector_test_case.py [hunk 11]
```diff
@@ -75,10 +79,14 @@ def __get_response(json_return_value, json_side_effect, content, text, headers,
...
     @staticmethod
-    def __post_response(json_return_value) -> AsyncMock:
+    def __post_response(json_return_value, json_side_effect: list = None) -> AsyncMock:
         """Create the mock post response."""
         post_response = AsyncMock()
-        post_response.json.return_value = json_return_value
+        if json_side_effect:
+            if not json_return_value:  # put the last side effect into return value, if it was not already given
+                json_return_value = json_side_effect[-1]
+            json_side_effect = json_side_effect + [STOP_SENTINEL]  # AsyncMock apparently does catch StopAsyncIteration
```
Comment (fniessink) [inline components/collector/tests/source_collectors/source_collector_test_case.py]: `json_side_effect` is a list, right? Why not:

```suggestion
            json_side_effect.append(STOP_SENTINEL)  # AsyncMock apparently does catch StopAsyncIteration
```

File: components/shared_data_model/src/shared_data_model/sources/azure_devops.py [hunk 12]
```diff
@@ -167,6 +177,10 @@
     entities=dict(
         failed_jobs=dict(name="failed pipeline", attributes=PIPELINE_ATTRIBUTES),
         job_runs_within_time_period=dict(name="pipeline", attributes=PIPELINE_ATTRIBUTES),
+        lead_time_for_changes=dict(
+            name="lead time",
+            attributes=ISSUE_ATTRIBUTES + [dict(name="Issue lead time in days", key="lead_time")]  # type int / days?
```
Comment (fniessink) [inline components/shared_data_model/src/shared_data_model/sources/azure_devops.py]: Why does the comment have a question mark? Is there a todo?

File: docs/src/changelog.md [hunk 13]
```diff
@@ -39,6 +39,7 @@ If your currently installed *Quality-time* version is not v4.4.0, please read th
 
 - New metric "Job runs within time period", to keep track of deployment frequency in Jenkins, GitLab and/or Azure DevOps. Closes [#3498](https://github.com/ICTU/quality-time/issues/3498).
 - Retrieve the possible projects and issue types for issue trackers from the configured issue tracker and present them via dropdown menus so the user can't enter non-existing projects or issue types. Closes [#4586](https://github.com/ICTU/quality-time/issues/4586).
+- New metric "Lead Time for Changes", to keep track of the average lead time of changes completed in a certain time period in Jira and/or Azure DevOps. Closes [#3501](https://github.com/ICTU/quality-time/issues/3501).
```
Comment (fniessink) [inline docs/src/changelog.md]: The entry is in the wrong location. Release 4.5.0 was already released. Please put it under the `v4.6.0-rc.1` header.

```suggestion
- New metric "Lead time for changes", to keep track of the average lead time of changes completed in a certain time period in Jira and/or Azure DevOps. Closes [#3501](https://github.com/ICTU/quality-time/issues/3501).
```

**Review by fniessink — CHANGES_REQUESTED** [review]

File: docs/src/changelog.md (hunk 13 above)
Comment (wkoot) [reply docs/src/changelog.md]: Ah yes the wonders of rebasing

File: components/shared_data_model/src/shared_data_model/sources/azure_devops.py (hunk 12 above)
Comment (wkoot) [reply components/shared_data_model/src/shared_data_model/sources/azure_devops.py]: Indeed it's a todo I forgot to mark as such. I wanted to look up the types, but I don't think it made much sense to change from default

File: components/collector/src/source_collectors/azure_devops/lead_time_for_changes.py (hunk 4 above)
Comment (wkoot) [reply components/collector/src/source_collectors/azure_devops/lead_time_for_changes.py]: Old habits for big data analysis code die hard :))

File: components/collector/src/source_collectors/azure_devops/user_story_points.py (hunk 7 above)
Comment (wkoot) [reply components/collector/src/source_collectors/azure_devops/user_story_points.py]: Yeah and similarly for Issues.. I've been thinking of extracting it, but `_parse_entities` is generic. Will change

File: components/collector/src/source_collectors/jira/lead_time_for_changes.py (hunk 9 above)
Comment (wkoot) [reply components/collector/src/source_collectors/jira/lead_time_for_changes.py]: Sadly, mean() does not work on an empty list :-)

File: components/collector/src/source_collectors/azure_devops/issues.py [hunk 14]
```diff
@@ -15,46 +16,84 @@ class AzureDevopsIssues(SourceCollector):
...
+        if not self._issue_ids_to_fetch:
             return SourceResponses(responses=[response], api_url=api_url)
-        ids_string = ",".join(ids[: min(self.MAX_IDS_PER_WORK_ITEMS_API_CALL, SourceMeasurement.MAX_ENTITIES)])
-        work_items_url = URL(f"{await super()._api_url()}/_apis/wit/workitems?ids={ids_string}&api-version=4.1")
-        work_items = await super()._get_source_responses(work_items_url, **kwargs)
+        work_items = await self._get_work_item_responses(auth)
         work_items.insert(0, response)
         return work_items
 
+    def _parse_entity(self, work_item: dict) -> dict:  # skipcq: PYL-R0201
```
Comment (fniessink) [inline components/collector/src/source_collectors/azure_devops/issues.py]: I understand why you decided not to have _parse_entity return an Entity, but since Entity is just thin wrapper around dict you could just as well return an Entity here and then add the extra attributes in the subclass with `entity[another_key] = value`. That would prevent the surprise that _parse_entity does not return an Entity.

File: components/collector/src/source_collectors/jira/lead_time_for_changes.py [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+        entity = super()._create_entity(issue, url)
+        entity["lead_time"] = self.__lead_time(issue["fields"])
+        return entity
+
+    def _compute_value(self, entities: Entities) -> Value:
+        """Calculate the average lead time of the completed issues."""
+        lead_times = []
+        for issue in entities:
+            lead_times.append(self.__lead_time(issue))
+        return str(round(mean(lead_times))) if entities else None
```
Comment (fniessink) [inline components/collector/src/source_collectors/jira/lead_time_for_changes.py]: I don't understand why you changed `_compute_value` from a class method to an instance method? It doesn't use any instance attributes?

File: components/collector/tests/source_collectors/source_collector_test_case.py [hunk 16]
```diff
@@ -75,10 +78,14 @@ def __get_response(json_return_value, json_side_effect, content, text, headers,
...
 
     @staticmethod
-    def __post_response(json_return_value) -> AsyncMock:
+    def __post_response(json_return_value, json_side_effect: list = None) -> AsyncMock:
         """Create the mock post response."""
         post_response = AsyncMock()
-        post_response.json.return_value = json_return_value
+        if json_side_effect:
+            if not json_return_value:  # put the last side effect into return value, if it was not already given
+                json_return_value = json_side_effect[-1]
```
Comment (fniessink) [inline components/collector/tests/source_collectors/source_collector_test_case.py]: The comment tells me what the code does, but I can read that for myself. It however doesn't explain *why* it does that.

File: components/collector/tests/source_collectors/source_collector_test_case.py [hunk 17]
```diff
@@ -75,10 +78,14 @@ def __get_response(json_return_value, json_side_effect, content, text, headers,
...
+            json_side_effect.append(STOP_SENTINEL)  # AsyncMock apparently does catch StopAsyncIteration
```
Comment (fniessink) [inline components/collector/tests/source_collectors/source_collector_test_case.py]: What's not apparent here is what happens if we don't append STOP_SENTINEL. Why is adding STOP_SENTINEL necessary?

**Review by fniessink — APPROVED** [review]

File: components/collector/src/source_collectors/jira/lead_time_for_changes.py (hunk 15 above)
Comment (wkoot) [reply components/collector/src/source_collectors/jira/lead_time_for_changes.py]: Hmm, I probably auto-refactored all the other ones when creating this method as an instance method.
Does it make sense to also refactor the other instance methods to classmethods, as we almost never use instance attributes?

File: components/collector/tests/source_collectors/source_collector_test_case.py (hunk 17 above)
Comment (wkoot) [reply components/collector/tests/source_collectors/source_collector_test_case.py]: We tried to debug this for a while, but that did not lead to an understanding :-) I'll update the comment

File: components/collector/src/source_collectors/jira/lead_time_for_changes.py (hunk 15 above)
Comment (fniessink) [reply components/collector/src/source_collectors/jira/lead_time_for_changes.py]: No, that's ok. It's not a big deal, but I was just wondering what I missed :-)
